lib/fizzy/parser_expr.cpp

```diff
@@ -155,34 +155,27 @@ inline uint8_t get_branch_arity(const ControlFrame& frame) noexcept
     return get_branch_frame_type(frame).has_value() ? 1 : 0;
 }
 
-inline void validate_branch_stack_height(const ControlFrame& current_frame,
-    const ControlFrame& branch_frame, const Stack<ValType>& operand_stack)
+inline void update_branch_stack(const ControlFrame& current_frame, const ControlFrame& branch_frame,
+    Stack<ValType>& operand_stack)
 {
-    const auto current_frame_stack_height = static_cast<int>(operand_stack.size());
+    assert(static_cast<int>(operand_stack.size()) >= current_frame.parent_stack_height);
 
-    assert(current_frame_stack_height >= current_frame.parent_stack_height);
-
-    const auto arity = get_branch_arity(branch_frame);
-    if (!current_frame.unreachable &&
-        (current_frame_stack_height < current_frame.parent_stack_height + arity))
-        throw validation_error{"branch stack underflow"};
+    const auto branch_frame_type = get_branch_frame_type(branch_frame);
+    if (branch_frame_type.has_value())
+        drop_operand(current_frame, operand_stack, *branch_frame_type);
 }
 
-
 void push_branch_immediates(const ControlFrame& branch_frame, int stack_height, bytes& immediates)
 {
-    const auto arity = get_branch_arity(branch_frame);
-
     // How many stack items to drop when taking the branch.
-    // This value can be negative for unreachable instructions.
-    const auto stack_drop = stack_height - branch_frame.parent_stack_height - arity;
+    const auto stack_drop = stack_height - branch_frame.parent_stack_height;
```

Comment:
It's not needed to subtract arity now, because now we pop the result from stack before calling this function (in `update_branch_stack`)